guardian/pluto-headers
# Feature/azure compatibility

## What does this change?

Adds the ability to authenticate with Azure AD via JKWS validation

This involves changing the authentication library to `jose`, which is no bad thing IMO but it does require upstream changes in clients.  Specifically, when using this version you need to _remove_ any dependencies on `jsonwebtoken` and _add_ a dependency on `jose` to the same version that the library expects.

## How to test

Build the library locally (following readme instructions) and link that into pluto-start (following readme instructions).  Deploy your modified pluto-start alongside an updated configuration:

Get hold of the SSO configuration from Azure (or other endpoint provider) like this:
```bash
curl https://login.microsoftonline.com/azure-tenant-id/v2.0/.well-known/openid-configuration | jq .
```

You'll need to use some of the values here to configure oauth-secrets.

Update the `config.json` file in the `oauth-secrets` manifest like this:
```json
{
  "clientId": "client-id-from-azure",
  "resource": "",
  "scope": "https://graph.microsoft.com/profile https://graph.microsoft.com/email https://graph.microsoft.com/user.read",
  "oAuthUri": "authorization_endpoint-from-sso-configuration",
  "tokenUri": "token_endpoint-from-sso-configuration",
  "jwksUri": "https://prexit.local/jwks/jwks.json",
  "jwksOriginUri": "jwks_uri-from-sso-configuration",
  "adminClaimName": "admin-role-name-from-azure-config"
}
```

`publickey.pem` is **no longer required** and will be ignored if `jwksUri` is set.
`jwksOriginUri` is ignored by pluto-headers but used by the `pluto-start` cronjob that syncs jwks certificates.

**NOTE** This configuration assumes that you are using the new cronjob in `pluto-start` to sync the jwks to https://prexit.local/jwks/jwks.json. For testing, you can set this URI the same as `jwksOriginUri` - this will fetch them from the orign every time. Not recommended for production but fine for testing.

Finally, **make sure pluto-start is restarted to get the new configurations**.

When you go to the pluto landing page and click "log me in", you should be directed out to Azure and then arrive back logged in.
So long as you have configured a "role" for administrator in the Azure configuration and set it up as per Rich's instructions you should be able to see admin/non-admin in action too.

Switching back is as simple as replacing the original `config.json` in `oauth-secrets.yaml`, deploying it and restarting pluto-start.  You should not have any problem running the updated code against the old authentication, it should be fully backwards compatible.

## How can we measure success?

Simple migration path to Azure

## Have we considered potential risks?

- Config switch should be done "out of hours" to minimise the risk of "breaking" user sessions
- If users get problems they should clear browser cache.  The config and public keys get cached by the local browser.
- Need to triple-check that there are no issues with rolling back and forth between azure and adfs
